crates/outbound-mqtt/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+    async fn open(
+        &mut self,
+        address: String,
+        username: String,
+        password: String,
+        keep_alive_interval: u64,
+    ) -> Result<Result<Resource<MqttConnection>, Error>> {
+        if !self.is_address_allowed(&address) {
+            return Ok(Err(Error::InvalidAddress));
+        }
+        self.establish_connection(
+            address,
+            username,
+            password,
+            Duration::from_secs(keep_alive_interval),
+        )
+        .await
+    }
+
+    async fn publish(
+        &mut self,
+        connection: Resource<MqttConnection>,
+        topic: String,
+        payload: Vec<u8>,
+        qos: Qos,
+    ) -> Result<Result<(), Error>> {
+        Ok(async {
+            let client = self.get_conn(connection).await.map_err(other_error)?;
+            let message = paho_mqtt::Message::new(&topic, payload, qos as i32);
+            client.publish(message).map_err(other_error)?;
+            Ok(())
+        }
+        .await)
+    }
+
+    fn drop(&mut self, connection: Resource<MqttConnection>) -> anyhow::Result<()> {
+        self.connections.remove(connection.rep());
+        Ok(())
+    }
+}
+
+fn other_error(e: impl std::fmt::Display) -> Error {
+    Error::Other(e.to_string())
+}
+
+/// Delegate a function call to the v2::HostConnection implementation
+macro_rules! delegate {
+    ($self:ident.$name:ident($address:expr, $username:expr, $password:expr, $dur:expr, $($arg:expr),*)) => {{
+        if !$self.is_address_allowed(&$address) {
+            return Ok(Err(v1::Error::Error));
+
+        }
+        let connection = match $self.establish_connection($address, $username, $password, $dur).await? {
+            Ok(c) => c,
+            Err(_) => return Ok(Err(v1::Error::Error)),
+        };
+        Ok(<Self as v2::HostConnection>::$name($self, connection, $($arg),*)
+            .await?
+            .map_err(|_| v1::Error::Error))
+    }};
+}
+
+#[async_trait]
+impl v1::Host for OutboundMqtt {
+    async fn publish(
+        &mut self,
+        address: String,
+        username: String,
+        password: String,
+        keep_alive_interval: u64,
+        topic: String,
+        payload: Vec<u8>,
+        qos: v1::Qos,
+    ) -> Result<Result<(), v1::Error>> {
+        delegate!(self.publish(
+            address,
+            username,
+            password,
+            Duration::from_secs(keep_alive_interval),
+            topic,
+            payload,
+            match qos {
+                v1::Qos::AtMostOnce => Qos::AtMostOnce,
+                v1::Qos::AtLeastOnce => Qos::AtLeastOnce,
+                v1::Qos::ExactlyOnce => Qos::ExactlyOnce,
+            }
+        ))
+    }
+}
+
+impl OutboundMqtt {
+    async fn get_conn(
+        &mut self,
+        connection: Resource<MqttConnection>,
+    ) -> Result<&mut Client, Error> {
+        self.connections
+            .get_mut(connection.rep())
+            .ok_or(Error::Other(
```

Comment:
In other places we use a ConnectionFailed variant for this case.  Is that not appropriate here?

Reply:
Was trying to be consistent with Redis implementation which uses `Other` as well? [ here ](https://github.com/fermyon/spin/blob/3ee465e8ae114d627a1441327bca44d5c9ad0084/crates/outbound-redis/src/lib.rs#L331) or did I misunderstand your comment?